### PR TITLE
Feature/link against newer libtiff

### PR DESCRIPTION
**Pull request by ignogueiras**
* Replace deprecated typedefs:    
      Since LibTiff v4.3.0 the use of "int8", "uint8", "int16", "uint16", "int32",
      "uint32", "int64", "uint64" typedefs is now deprecated. libtiff code and
      headers no longer use them, and use their C99 standard equivalents
      (with "_t" suffix). Right now this only causes deprecation warning messages
      to appear at compile time, but the typedefs will be removed in a future release.
* Link against libtiffxx:
      Since v3.7.2 LibTiff place the C++ stream API in the separate library called
      libtiffxx to avoid unneeded dependencies. This change generates two separated .so
      libraries, libtiff.so and libtiffxx.so. PDFWriter was linking against libtiff.so alone,
      preventing me from using functions that receive C++ streams as parameters.
      With this change the library should link agains both .so.

With this modifications I can compile using new versions of LibTiff, while still compatible with your bundled version, so I think it should be a safe change. But let me know if you see any problem.

### CODE REVIEW

File: PDFWriter/CMakeLists.txt [hunk 1]
```diff
@@ -405,7 +405,7 @@ if(PDFHUMMUS_NO_TIFF)
     target_compile_definitions(PDFWriter PUBLIC PDFHUMMUS_NO_TIFF=1)    
 else(PDFHUMMUS_NO_TIFF)
     if(USING_UNBUNDLED_TIFF)
-        target_link_libraries(PDFWriter TIFF::TIFF)
+        target_link_libraries(PDFWriter ${TIFF_LIBRARIES})
```
Comment (galkahana) [inline PDFWriter/CMakeLists.txt:408]: why is this change required? i'd rather use targets.

**Comment by galkahana** [comment]
Thanks. the changes in tiff handler seems good, but im not sure im on board with the changes in cmake. 
how about i prep an MR that upgrades the bundled tiff package and makes the necessary changes in tiffhandler like you proposed? maybe this will work to leave the cmake setup as is and make the bundled and unbundled versions compatible? i'll post here when i have it.

**Comment by galkahana** [comment]
sounds like we can use the libtiff:cxx target instead. like, do the find_package(tiff cxx) (drop the required) and then use the cxx target. b.t.w can hummus import one thing and your thing another? doesn't it work well togather. well...whatever, i'll do the thing and get back to you.

File: PDFWriter/CMakeLists.txt (hunk 1 above)
Comment (galkahana) [reply PDFWriter/CMakeLists.txt:408]: maybe even just add it in your own cmakefiles.

File: CMakeLists.txt [hunk 2]
```diff
@@ -92,7 +92,7 @@ if(NOT USE_BUNDLED)
 
     # libtiff
     if(NOT PDFHUMMUS_NO_TIFF)
-        find_package (TIFF)
+        find_package (TIFF REQUIRED CXX)
```
[the remark this answers is not in the document]
Comment (galkahana) [reply CMakeLists.txt:95]: so just find_package(TIFF CXX) (or some way to customize the input components. we will need CXX to allow a later CXX target).

File: PDFWriter/CMakeLists.txt (hunk 1 above)
Comment (galkahana) [reply PDFWriter/CMakeLists.txt:408]: yeah, i reckon you should be linking to the target directly. And i recommend using target_link_libraries(YOUR_LIBRARY TIFF:CXX) in addition to  target_link_libraries(YOUR_LIBRARY TIFF:TIFF) to get both the tiff target and C++ wrapper.

It does still mean that find_package should bring in the CXX component.
not sure if you can do this by just adding find_package(TIFF COMPONETS CXX) in your cmake file. im not sure how multiple find_package work...but it might work. if not, try adding CXX as you did to the pdfwriter as you did (though maybe without the "required" bit).

in any case, i'll be working on an upgrade to the bundled libtiff using the source code changes you made, so at least this part is a bit more compatible to the new version.

Comment (galkahana) [reply PDFWriter/CMakeLists.txt:408]: given that PDFWriter itself does not use the C++ wrapper, there's no reason for it to look or link against it, unless there's no other way to overcome this, one was suggested here.

**Comment by galkahana** [comment]
So, this MR - https://github.com/galkahana/PDF-Writer/pull/227 - upgrades the bundled tiff to 4.6. this should take care of the first issue you raised. as for the second one i recommend an independent import in your project, as pdfhummus itself uses the C interface...so it doesn't make much sense it'll compile+link against the C++ target. I think this should be possible as i specified. do let me know how it works for you, if you try it.

**Comment by ignogueiras** [comment]
It just felt a little strange to me to link against the same lib several times from several targets in the chain, but as you said, it should be the right choice here.
I just tested it, reverted the changes in the CMake files in PDFWriter and linked LibTiff directly from the top level app.
As you suggested I added `find_package(TIFF COMPONENTS CXX)` in our CMakes and linked `TIFF::CXX`.
It works perfectly this way, so thank you very much for your insights and for dealing with my mistakes.

**Comment by galkahana** [comment]
Thanks for testing an letting me know how it worked. I'm glad this works out for you.

**Comment by galkahana** [comment]
Hummus v4.6 now includes an upgraded libtiff which should take care of potential future deprecation as you mentioned. Thanks.

**Comment by galkahana** [comment]
I think we can close this? if not please reopen.
